torch/csrc/Module.cpp

```diff
@@ -267,6 +265,34 @@ PyObject *THPModule_addDocStr(PyObject *_unused, PyObject *args)
           "Type '%s' already has a docstring", t->tp_name);
     }
     t->tp_doc = doc_str;
+  // Pybind11 method
+  } else if (Py_TYPE(obj) == &PyInstanceMethod_Type) {
+    PyObject* func_obj = PyInstanceMethod_GET_FUNCTION(obj);
+    if (Py_TYPE(func_obj) != &PyCFunction_Type) {
+      return PyErr_Format(PyExc_TypeError,
+          "don't know how to add docstring for instancemethod with non-C implementation");
+    }
+    PyCFunctionObject* func = (PyCFunctionObject*)func_obj;
+    // Remove pybind11's default docstring
+    if (func->m_ml->ml_doc) {
+      // We expect pybind11's automatic docstring generation to be disabled
+      // In this case, pybind11 currently produces an empty docstring for the function
+      if (strlen(func->m_ml->ml_doc) != 0) {
+        return PyErr_Format(PyExc_RuntimeError,
+            "pybind11 method '%s' already has a docstring "
+            "(to fix this error, disable pybind11's automatic docstring generation)",
+            func->m_ml->ml_name);
+      }
+      // This is extremely hacky; it uses the implementation detail of pybind11 that
+      // docstrings are allocated with `malloc()`
+      //NOLINTNEXTLINE(cppcoreguidelines-no-malloc, cppcoreguidelines-pro-type-const-cast)
+      free(const_cast<char *>(func->m_ml->ml_doc));
```

Comment:
@ezyang WDYT? I need second opinion

Reply:
@pbelevich I actually submitted a PR to pybind11 a few weeks ago to make pybind11 not assign an empty docstring when all docstring-related options are disabled, however even with that change the `strdup` below is still needed, let alone that change will need time to make its way to pybind11's next release and FB's internal repo...

Reply:
We bundle pybind11, so we can take these changes immediately. If it substantially simplifies the implementation in this PR (which I think it does), I'd prefer if we first do a submodule update of pybind11 to the latest master, and then use that to base this PR off of. Don't use ghstack for that, ghstack can't do submodule updates. Someone on FB will have to spearhead the process; @pbelevich, do you think you'll have time to do it? Also cc @wanchaol @seemethere who did updates in the recent history.

Reply:
@ezyang FYI that PR is pybind/pybind11#2745. With that change we should be able to get rid of these four lines. However, line 291-295 is still necessary because pybind11 manages docstrings with `malloc` and `free`, while `doc_str` (backed by a `std::string`) is probably allocated with operator `new`. If we assign `doc_str` to `ml_doc`, the program may crash when the function is being destructed because it's trying to free the memory of `doc_str` in an incompatible way.

Reply:
cc @malfet